src/trees/NodeIndex.h

```diff
@@ -33,117 +33,78 @@
 
 #include <iostream>
 
-#include "MRCPP/mrcpp_declarations.h"
-
 namespace mrcpp {
 
 template <int D> class NodeIndex final {
 public:
-    NodeIndex(int n = 0, const int *l = nullptr);
-    NodeIndex(const NodeIndex<D> &idx);
-    NodeIndex(const NodeIndex<D> &pIdx, int cIdx);
+    // regular constructors
+    NodeIndex(int n = 0, const std::array<int, D> &l = {})
+            : N(static_cast<short int>(n))
+            , L(l) {}
+
+    // relative constructors
+    NodeIndex<D> parent() const {
+        std::array<int, D> l;
+        for (int d = 0; d < D; d++) l[d] = (this->L[d] < 0) ? (this->L[d] - 1) / 2 : this->L[d] / 2;
+        return NodeIndex<D>(this->N - 1, l);
+    }
+    NodeIndex<D> child(int cIdx) const {
+        std::array<int, D> l;
+        for (int d = 0; d < D; d++) l[d] = (2 * this->L[d]) + ((cIdx >> d) & 1);
+        return NodeIndex<D>(this->N + 1, l);
+    }
 
-    inline NodeIndex<D> &operator=(const NodeIndex<D> &idx);
-    inline bool operator==(const NodeIndex<D> &idx) const;
-    inline bool operator!=(const NodeIndex<D> &idx) const;
+    // comparisons
+    bool operator!=(const NodeIndex<D> &idx) const { return not(*this == idx); }
+    bool operator==(const NodeIndex<D> &idx) const {
+        bool out = (this->N == idx.N) ? true : false;
```

Comment:
I think you can skip the ternary if here, `this-> N == idx.N` already gives `true` or `false`.